Site-creation rework: Site creation domain screen

hypest opened the pull request:
Fixes #6928 

Known issue: the suggestions coming out of the FluxC domain suggestions api call don't seem to respect the wpcom-subdomains-only flag. This will be handled in a separate PR.

Implementation info:
* Similar to the headless "loader" Fragment in the themes selection screen, a headless fragment is employed here too (`SiteCreationDomainLoaderFragment`) for handling the communication with FluxC in a rotation-friendly manner, without losing the result of the network call if that happens to arrive during a device rotation.
* The whole UI is implemented as content of a RecyclerView so scrolling can be easy and has enough touch height to be performed when device height is small (think: landscape orientation). The textual header ("Step 4 of 4..") could be static and non-scrolling but in this first implementation I opted to keeping it in for maximum scrollable height.

To test:
* With the app logged in wpcom, create a new wpcom site via the site picker
* Reach the domain selection screen
* Notice the progress indicator whenever there is an active query in progress
* Enter a query text to trigger a new query and notice the list refreshing after a while
* Hit the "Create site" button to submit the selection. A Toast should pop up with the selected domain name.

theck13 commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SiteCreationDomainAdapter.java:
```diff
@@ -0,0 +1,199 @@
+                            public void run() {
+                                mOnAdapterListener.onKeywordsChange(editable.toString());
+                            }
+                        }, 400, TimeUnit.MILLISECONDS);
```
We should make this a constant to help us use the same value for similar interfaces across the app.  As an example, [the Username Changer screen uses `2000` milliseconds](https://github.com/wordpress-mobile/WordPress-Android/pull/7151/files#diff-4e6863adecd7157709266b8ea7819618R72).

hypest replied:
Done with 7741abc.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SiteCreationDomainAdapter.java:
```diff
@@ -0,0 +1,199 @@
+
+    @Override
+    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+        if (viewType == VIEW_TYPE_HEADER) {
```
Done with a209b46.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SiteCreationDomainAdapter.java:
```diff
@@ -0,0 +1,199 @@
+    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
+        int viewType = getItemViewType(position);
+
+        if (viewType == VIEW_TYPE_HEADER) {
```
Done with 5052a9f.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SiteCreationDomainAdapter.java:
```diff
@@ -0,0 +1,199 @@
+        int viewType = getItemViewType(position);
+
+        if (viewType == VIEW_TYPE_HEADER) {
+//            final HeaderViewHolder headerViewHolder = (HeaderViewHolder) holder;
```
Was indeed intentionally commented out but I never went back to add more info. Fixed that in https://github.com/wordpress-mobile/WordPress-Android/commit/5052a9f932d434ad87443db7c0dbb3103aee4826#diff-12fe24c901e61f13416d3cbfab63f6e4R169

hypest commented:
Thanks for the detailed review @theck13 !

> I like the progress bar overlaying the search icon when fetching domain suggestions. I think I'll mimic that in the Username Changer implementation.

Nice!

>  I noticed the progress bar can remain invisible while searching in certain circumstances.

Indeed! Fixed with 3bc8a0b.

> The selected domain is not maintained on device rotation.

Indeed! Fixed with 05000a6. While at it, I also updated the enabled/disabled behavior of the "Create site" button to be disabled if nothing is selected.

> The text input view does not show any suggestions when entered text is deleted

Indeed! Let's show the suggestions based on the username. Fixed with 661457e.

> If a search is initiated after selecting a suggestion form the list, the selected suggestion is not maintained and a new suggestion is selected.

That's the intended behavior implemented. Selection always happens from the new results. Does this miss some design specification?

> The text input view loses focus when scrolled off screen requiring the user to tap the view again to continue searching for suggestions.

Indeed! Fixed with 79de41c.

> I'm wondering if the support library dependency could cause bugs down the road with multiple view types and varying layouts.

I'd assume this is improbable since the RecycleView Adapter is [designed to support](https://developer.android.com/reference/android/support/v7/widget/RecyclerView.Adapter.html#onCreateViewHolder(android.view.ViewGroup,%20int)) exactly that (heterogeneous view types).

> ... there were two suggestions on screen as shown in the screenshot below, which seemed like enough to me.

Two suggestions may be enough, but the problem the PR's solution attacks is that the _scrollable height_ is too small. Trying to scroll such a small container is really awkward and effectively, a "broken window".

> The Domain Selection screen is very similar to the Username Changer screen. We should share code between the two if possible.

Hopefully we can do that down the road. At the moment, out of coincidence it happened that the username changer and the domain suggestion screens were developed in parallel. Perhaps we can revisit and make an attempt to converge the solutions there.

Ready for another pass! Thanks!

theck13 commented:
Nice work addressing my comments!

<blockquote><blockquote>If a search is initiated after selecting a suggestion form the list, the selected suggestion is not maintained and a new suggestion is selected.</blockquote>

That's the intended behavior implemented. Selection always happens from the new results. Does this miss some design specification?</blockquote>

I don't see anything specified in the Domain Selection design.  The behavior was discussed (quoted below) regarding the Username Changer design, which is very similar to the Domain Selection design so we may want to adopt the same behavior for consistency.

<blockquote><blockquote>Will the newly selected username be moved to the top of the list?</blockquote>

No, unless the user triggers a new search, in which case it will move at the top, and the other results below.</blockquote>

folletto commented:
> The textual header ("Step 4 of 4..") could be static and non-scrolling but in this first implementation I opted to keeping it in for maximum scrollable height.

Yep. The whole page should scroll, intro text included.

> The text input view does not show any suggestions when entered text is deleted. Should we show a dialog explaining why no suggestions are displayed, display the suggestions from the initial screen loading, or keep the previous suggestions?

Yep, thanks for the note and the fix. In general, the list should strive to be NEVER empty.

> Should the Domain Selection and Username Changer screens be similar or distinctly different?

Yep they are made to be different, as they come from two different goals and copy. Let's keep the as they are for now.

> No, unless the user triggers a new search, in which case it will move at the top, and the other results below.

Yep, correct. The selected option should always be preserved.

Also, we shouldn't pre-select an option for them as I noticed from the GIF above.

hypest commented:
> > No, unless the user triggers a new search, in which case it will move at the top, and the other results below.
> 
> Yep, correct. The selected option should always be preserved.
>
> Also, we shouldn't pre-select an option for them as I noticed from the GIF above.

Cool, will work on both @folletto , thanks for confirming!

hypest commented:
> There is some weird behavior while typing, which might be a result of the focus changes in 79de41c. In the animation below, I typed "testing" then cleared the text.

Indeed, fixed with d552c0e.

hypest commented:
> The selected option should always be preserved.

Done with 3fa4f90 .

> Also, we shouldn't pre-select an option for them as I noticed from the GIF above.

Done with 594a6be.

hypest commented:
Ready for another pass @theck13 , thanks!

theck13 commented on WordPress/src/main/java/org/wordpress/android/ui/accounts/signup/SiteCreationDomainAdapter.java:
```diff
@@ -0,0 +1,277 @@
+
+    private String mCarryOverDomain;
+    private String mSelectedDomain;
+    private boolean mNeedExtraLineForSelectedDomain;
```
I don't follow the difference between `mCarryOverDomain` and `mSelectedDomain`.  When are they different?  Also, wouldn't `mNeedExtraLineForSelectedDomain` be the same as if `mSelectedDomain` has no value?

hypest replied:
`Carry-over` and `selected` can be different when the user is on a screen where there is the top-item that got carried over from the previous query and then the user has selected a different item, from the new query. So, on screen, there will be both the carry-over (at the top) and the selected (somewhere down the list).

So, need to keep them in separate vars so when the list refreshes (happens when selecting another list item for instance) the adapter knows what to render and where.

Let me know if that makes sense @theck13 , thanks!
